aiothrift/protocol.py

```diff
@@ -370,14 +372,65 @@ async def skip(reader, ftype):
             await skip(reader, f_type)
 
 
+class TFramedTransport:
+    """Implement the Twisted framed transport protocol.
+
+    Since most async servers use this including the python2 twisted
+    bindings, this is likely of interest.
+    """
+    def __init__(self, base):
+        self.__base = base
+        self.__read_buffer = BytesIO()
+        self.__write_buffer = BytesIO()
+
+    def read(self, n=-1):
+        if len(self.__read_buffer.getvalue()) == 0:
+            self.readFrame()
+        return self.__read_buffer.read(n)
+
+    async def readFrame(self):
+        buff = await self.__base.readexactly(4)
+        sz, = unpack('!i', buff)
+        self.__read_buffer = BytesIO(await self.__base.readexactly(sz))
+
+    async def readexactly(self, n):
+        now, end = self.__read_buffer.tell(), self.__read_buffer.seek(0, 2)
+        remaining = end - now
+        self.__read_buffer.seek(now)
+
+        if 0 < remaining < n:
+            raise IOError("Tried to read invalid amount from framed transport")
```

Comment:
How can we ensure that `remaining >= n`?
Would it better to firstly `await readframe` and `await readexactly` again instead of just
raising error?

Reply:
I think it will always be an error to hit this (at least from the point of view of this library).  Basically the way this works is when you call read() it reads the entire frame into __read_buffer.  As long as a call to readexactly() doesn't cross frame boundaries this works (although if you do it will error, I agree).

The question is if we expect this transport to be called generically by some other caller or only by the other thrift code in this library.  Since due to the way the rest of the thrift stack here works it will never make a call that crosses the transport frame boundaries.

So we could await here to read more if that would be less confusing, but imo it is probably a bug in the caller's code if it ever does that.

Reply:
Ok，this seems reasonable. Thank u.